File: iosApp/iosApp/Firebase/FirebaseAuth.swift
import FirebaseAuth

final class FirebaseAuth {
    private static let auth = Auth.auth()
    
    func currentUser() -> User? { return FirebaseAuth.auth.currentUser }
    func reload(callback: @escaping (AuthErrorCode?) -> Void) {
        currentUser()?.reload { error in
            if let error = error as? NSError {
                callback(AuthErrorCode(rawValue: error.code))
            }
            else {
                callback(nil)
            }
        }
    }
    func isEmailVerified() -> Bool? { currentUser()?.isEmailVerified }
    func createAccount(_ emailAddress: String, _ password: String, callback: @escaping (AuthErrorCode?) -> Void) {
        FirebaseAuth.auth.createUser(withEmail: emailAddress, password: password) { result, error in
            if let error = error as? NSError {
                callback(AuthErrorCode(rawValue: error.code))
            }
            else {
                self.reload { reloadError in
                    callback(reloadError)
                }
            }
        }
    }
    func signIn(_ emailAddress: String, _ password: String, callback: @escaping (AuthErrorCode?) -> Void) {
        FirebaseAuth.auth.signIn(withEmail: emailAddress, password: password) { result, error in
            if let error = error as? NSError {
                callback(AuthErrorCode(rawValue: error.code))
            }
            else {
                self.reload { reloadResult in
                    callback(reloadResult)
                }
            }
        }
    }
    func sendPasswordReset(_ emailAddress: String, callback: @escaping (AuthErrorCode?) -> Void) {
        FirebaseAuth.auth.sendPasswordReset(withEmail: emailAddress) { error in
            if let error = error as? NSError {
                callback(AuthErrorCode(rawValue: error.code))
            }
            else {
                callback(nil)
            }
        }
    }
    func sendEmailVerification(callback: @escaping (AuthErrorCode?) -> Void) {
        currentUser()?.sendEmailVerification { error in
            if let error = error as? NSError {
                callback(AuthErrorCode(rawValue: error.code))
            }
            else {
                callback(nil)
            }
        }
    }
    func updateEmailAddress(_ emailAddress: String, callback: @escaping (AuthErrorCode?) -> Void) {
        currentUser()?.sendEmailVerification(beforeUpdatingEmail: emailAddress) { error in
            if let error = error as? NSError {
                callback(AuthErrorCode(rawValue: error.code))
            }
            else {
                callback(nil)
            }
        }
    }
    func signOut() -> Bool {
        do {
            try FirebaseAuth.auth.signOut()
            return true
        }
        catch {
            return false
        }
    }
}

File: androidApp/src/main/java/com/cistus/hunter/android/firebase/FirebaseAuth.kt
package com.cistus.hunter.android.firebase

import com.google.firebase.auth.AuthCredential
import com.google.firebase.auth.FirebaseAuthException
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.ktx.auth
import com.google.firebase.ktx.Firebase

class FirebaseAuth {
    companion object {
        private val auth = Firebase.auth
    }

    fun currentUser(): FirebaseUser? = auth.currentUser
    fun reload(callback: (String?) -> Unit) {
        currentUser()
            ?.reload()
            ?.addOnSuccessListener { callback(null) }
            ?.addOnFailureListener {
                if (it is FirebaseAuthException)
                    callback(it.errorCode)
                else
                    callback(it.localizedMessage)
            }
    }
    fun reAuthenticate(authCredential: AuthCredential, callback: (String?) -> Unit) {
        currentUser()?.let { currentUser ->
            currentUser.reauthenticate(authCredential)
                .addOnSuccessListener { callback(null) }
                .addOnFailureListener {
                    if (it is FirebaseAuthException)
                        callback(it.errorCode)
                    else
                        callback(it.localizedMessage)
                }
        }
    }
    fun isEmailVerified(): Boolean? = currentUser()?.isEmailVerified
    fun createAccount(emailAddress: String, password: String, callback: (String?) -> Unit) {
        auth.createUserWithEmailAndPassword(emailAddress, password)
            .addOnSuccessListener {
                reload { reloadResult ->
                    callback(reloadResult)
                }
            }
            .addOnFailureListener {
                if (it is FirebaseAuthException)
                    callback(it.errorCode)
                else
                    callback(it.localizedMessage)
            }
    }
    fun signIn(emailAddress: String, password: String, callback: (String?) -> Unit) {
        auth.signInWithEmailAndPassword(emailAddress, password)
            .addOnSuccessListener {
                reload { reloadResult ->
                    callback(reloadResult)
                }
            }
            .addOnFailureListener {
                if (it is FirebaseAuthException)
                    callback(it.errorCode)
                else
                    callback(it.localizedMessage)
            }
    }
    fun sendPasswordReset(emailAddress: String, callback: (String?) -> Unit) {
        auth.sendPasswordResetEmail(emailAddress)
            .addOnSuccessListener { callback(null) }
            .addOnFailureListener {
                if (it is FirebaseAuthException)
                    callback(it.errorCode)
                else
                    callback(it.localizedMessage)
            }
    }
    fun sendEmailVerification(callback: (String?) -> Unit) {
        currentUser()
            ?.sendEmailVerification()
            ?.addOnSuccessListener { callback(null) }
            ?.addOnFailureListener {
                if (it is FirebaseAuthException)
                    callback(it.errorCode)
                else
                    callback(it.localizedMessage)
            }
    }
    fun updateEmailAddress(emailAddress: String, callback: (String?) -> Unit) {
        currentUser()
            ?.verifyBeforeUpdateEmail(emailAddress)
            ?.addOnSuccessListener { callback(null) }
            ?.addOnFailureListener {
                if (it is FirebaseAuthException)
                    callback(it.errorCode)
                else
                    callback(it.localizedMessage)
            }
    }
    fun signOut() = auth.signOut()

}